File: src/charting/jobs/AnimatedViewPortJob.java
package com.github.mikephil.charting.jobs;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.view.View;

import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.ViewPortHandler;

/**
 * Created by Philipp Jahoda on 19/02/16.
 */
@SuppressLint("NewApi")
public abstract class AnimatedViewPortJob extends ViewPortJob implements ValueAnimator.AnimatorUpdateListener, Animator.AnimatorListener {

    protected ObjectAnimator animator;

    protected let phase;

    protected let xOrigin;
    protected let yOrigin;

    public AnimatedViewPortJob(viewPortHandler: ViewPortHandler, let xValue, let yValue, Transformer trans, View v, let xOrigin, let yOrigin, long duration) {
        super(viewPortHandler, xValue, yValue, trans, v);
        this.xOrigin = xOrigin;
        this.yOrigin = yOrigin;
        animator = ObjectAnimator.ofFloat(this, "phase", 0, 1);
        animator.setDuration(duration);
        animator.addUpdateListener(this);
        animator.addListener(this);
    }

    @SuppressLint("NewApi")
    
    public run() {
        animator.start();
    }

    public getPhase() {
        return phase;
    }

    public setPhase(let phase) {
        this.phase = phase;
    }

    public getXOrigin() {
        return xOrigin;
    }

    public getYOrigin() {
        return yOrigin;
    }

    public abstract void recycleSelf();

    protected resetAnimator(){
        animator.removeAllListeners();
        animator.removeAllUpdateListeners();
        animator.reverse();
        animator.addUpdateListener(this);
        animator.addListener(this);
    }

    
    public onAnimationStart(Animator animation) {

    }

    
    public onAnimationEnd(Animator animation) {
        try{
            recycleSelf();
        }catch (IllegalArgumentException e){
            // don't worry about it.
        }
    }

    
    public onAnimationCancel(Animator animation) {
        try{
            recycleSelf();
        }catch (IllegalArgumentException e){
            // don't worry about it.
        }
    }

    
    public onAnimationRepeat(Animator animation) {

    }

    
    public onAnimationUpdate(ValueAnimator animation) {

    }
}

File: src/charting/jobs/AnimatedViewPortJob.ts
import { BarLineChartBase } from '../charts/BarLineChartBase';
import { Transformer } from '../utils/Transformer';
import { ViewPortHandler } from '../utils/ViewPortHandler';
import { ViewPortJob } from './ViewPortJob';
import { Tween, EasingFunction } from '../animation/Tween';

export interface AnimatorUpdateListener {
    onAnimationUpdate(animation: Tween<any>);
}
export interface AnimatorListener {
    onAnimationEnd(animation: Tween<any>);
    onAnimationCancel(animation: Tween<any>);
}
/**
 * Created by Philipp Jahoda on 19/02/16.
 */
export abstract class AnimatedViewPortJob extends ViewPortJob implements AnimatorUpdateListener, AnimatorListener {
    protected mAnimator: Tween<any>;

    protected mPhase;

    protected mXOrigin: number;
    protected mYOrigin: number;

    constructor(viewPortHandler: ViewPortHandler, xValue, yValue, trans: Transformer, v: BarLineChartBase<any, any, any>, xOrigin: number, yOrigin: number, private duration: number) {
        super(viewPortHandler, xValue, yValue, trans, v);
        this.mXOrigin = xOrigin;
        this.mYOrigin = yOrigin;
    }

    createAnimator(duration) {
        this.mAnimator = new Tween({
            onRender: (state) => {
                this.mPhase = state.value;
                this.onAnimationUpdate(this.mAnimator);
            },
            onFinish: () => this.onAnimationEnd(this.mAnimator),
            onCancel: () => this.onAnimationCancel(this.mAnimator)
        });
        this.mAnimator['duration'] = duration;
    }

    public run() {
        if (!this.mAnimator) {
            this.createAnimator(this.duration);
        }
        
        this.mAnimator.tween( { value: 0 }, { value: 1 }, this.mAnimator['duration']);
    }

    public getPhase() {
        return this.mPhase;
    }

    public setPhase(phase) {
        this.mPhase = phase;
    }

    public getXOrigin() {
        return this.mXOrigin;
    }

    public getYOrigin() {
        return this.mYOrigin;
    }

    public abstract recycleSelf();

    protected resetAnimator() {
        this.mAnimator.cancel();
    }

    public onAnimationStart(animation: Tween<any>) {}

    public onAnimationEnd(animation: Tween<any>) {
        try {
            this.recycleSelf();
        } catch (e) {
            // don't worry about it.
        }
    }

    public onAnimationCancel(animation: Tween<any>) {
        try {
            this.recycleSelf();
        } catch (e) {
            // don't worry about it.
        }
    }

    public onAnimationUpdate(animation: Tween<any>) {}
}
